include/net/downloader.h

```diff
@@ -0,0 +1,352 @@
+	DOWNLOADER_EVT_STOPPED,
+	/** Downloader deinitialized. Memory can be freed. */
+	DOWNLOADER_EVT_DEINITIALIZED,
+};
+
+struct download_fragment {
+	const void *buf;
+	size_t len;
+};
+
+/**
+ * @brief Downloader event.
+ */
+struct downloader_evt {
+	/** Event ID. */
+	enum downloader_evt_id id;
+
+	union {
+		/** Error cause. */
+		int error;
+		/** Fragment data. */
+		struct download_fragment fragment;
+	};
+};
+
+/**
+ * @brief Downloader asynchronous event handler.
+ *
+ * Through this callback, the application receives events, such as
+ * download of a fragment, download completion, or errors.
+ *
+ * On a @c DOWNLOADER_EVT_ERROR event with error @c ECONNRESET,
+ * returning zero from the callback will let the library attempt
+ * to reconnect to the server and continue the download.
+ * Otherwise, the callback may return any non-zero value
+ * to stop the download. On any other error code than @c ECONNRESET, the downloader
+ * will not attempt to reconnect and ignores the return value.
+ * To resume the download, use @ref downloader_get().
+ *
+ * @param[in] event	The event.
+ *
+ * @return Zero to continue the download, non-zero otherwise.
+ */
+typedef int (*downloader_callback_t)(const struct downloader_evt *event);
+
+/**
+ * @brief Downloader configuration options.
+ */
+struct downloader_cfg {
+	/** Event handler. */
+	downloader_callback_t callback;
+	/** Downloader buffer. */
+	char *buf;
+	/** Downloader buffer size. */
+	size_t buf_size;
+};
+
+/**
+ * @brief Downloader host configuration options.
+ */
+struct downloader_host_cfg {
+	/**
+	 * TLS security tag list.
+	 * Pass NULL to disable TLS.
+	 * The list must be kept in scope while download is going on.
+	 */
+	const int *sec_tag_list;
+	/**
+	 * Number of TLS security tags in list.
+	 * Set to 0 to disable TLS.
+	 */
+	uint8_t sec_tag_count;
+	/**
+	 * PDN ID to be used for the download.
+	 * Zero is the default PDN.
+	 */
+	uint8_t pdn_id;
+	/**
+	 * Range override.
+	 * Request a number of bytes from the server at a time.
+	 * 0 disables the range override, and the downloader will ask for the whole file.
+	 * The nRF91 series has a limitation of decoding ~2k of data at once when using TLS, hence
+	 * range override will be used in this case regardless of the value here.
+	 */
+	size_t range_override;
+	/** Use native TLS */
+	bool set_native_tls;
+	/**
+	 * Keep connection to server when done.
+	 * Server is disconnected if a file is requested from another server, and when the
+	 * downloader is deinitialized.
+	 */
+	bool keep_connection;
+	/**
+	 * Enable DTLS connection identifier (CID) feature.
+	 * This option require modem firmware version >= 1.3.5.
+	 */
+	bool cid;
+	/**
+	 * Address family to be used for the download, AF_INET6 or AF_INET.
+	 * Set to AF_UNSPEC (0) to fallback to AF_INET if AF_INET6 does not work.
+	 */
+	int family;
+};
+
+/**
+ * @brief Downloader internal state.
+ */
+enum downloader_state {
+	DOWNLOADER_DEINITIALIZED = 0x1,
```

Comment:
Because then we can expect multiple before states in `state_set()`

Reply:
Does it mean that it can be in multiple states at the same time?

Reply:
No. The idea was that for instance in downloader_startl we can be in both connected and idle state and would like to change to connecting. Though, I see it would be clearer to aboid it, so I will rewrite it.